### PR TITLE
Added implementation of LLM profile distillation to SASRec

### PR DESCRIPTION
**Pull request by Nikis14**

### CODE REVIEW

File: replay/models/nn/sequential/sasrec_with_llm/lightning.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        user_profile_emb: torch.Tensor,
+        hidden_for_reconstruction: torch.Tensor,
+        null_profile_binary_mask_batch: torch.BoolTensor,
+        criterion_reconstruct_fn
+    ) -> torch.Tensor:
+        user_profile_emb_transformed = self._model.aggregate_profile(user_profile_emb)
+        if self._model.use_upscale:
+            hidden_for_reconstruction = self._model.hidden_layer_transform(hidden_for_reconstruction)
+
+        null_profile_binary_mask_batch = null_profile_binary_mask_batch.flatten()
```
Comment (shashist) [inline replay/models/nn/sequential/sasrec_with_llm/lightning.py]: Seems like this line equalizes null profiles with corresponding hidden_for_reconstruction. 
It seems more convinient than assigning zero embedding, but why not skip null profiles in loss?

File: replay/models/nn/sequential/sasrec_with_llm/lightning.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        epoch = self.trainer.current_epoch
+        if self.scale_guide_loss:
+            loss_model_val = loss_model.item()
+            loss_guide_val = loss_guide.item()
+            eps = 1e-8
+
+            scale_for_guide = loss_model_val / (loss_guide_val + eps)
+            scaled_loss_guide = loss_guide * scale_for_guide
+
+            if epoch < self.profile_distil_epochs:
```
Comment (shashist) [inline replay/models/nn/sequential/sasrec_with_llm/lightning.py]: Could we move this check to `_compute_loss` to prevent unnecessary `loss_guide` computation?

File: replay/models/nn/sequential/sasrec_with_llm/utils.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+import random
+import numpy as np
+import torch
+import json
+
+from torch import nn
+import torch.nn.functional as F
+
+
+def set_seed(seed):
```
Comment (shashist) [inline replay/models/nn/sequential/sasrec_with_llm/utils.py]: Are functions `set_seed`, `init_criterion_reconstruct`, `calculate_recsys/guide_loss` used in the code?

File: replay/models/nn/sequential/sasrec_with_llm/dataset.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+class SasRecLLMTrainingDataset(SasRecTrainingDataset):
+    """
+    Dataset that generates samples to train SasRecLLM-like model
+    """
+
+    def __init__(
+        self,
+        sequential: SequentialDataset,
+        max_sequence_length: int,
+        user_profile_embeddings: torch.FloatTensor,
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec_with_llm/dataset.py:38]: Docstring for new arguments is missing

File: replay/models/nn/sequential/sasrec_with_llm/lightning.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+        dropout_rate: float = 0.2,
+        ti_modification: bool = False,
+        time_span: int = 256,
+        loss_type: str = "CE",
+        loss_sample_count: Optional[int] = None,
+        negative_sampling_strategy: str = "global_uniform",
+        negatives_sharing: bool = False,
+        optimizer_factory: OptimizerFactory = FatOptimizerFactory(),
+        lr_scheduler_factory: Optional[LRSchedulerFactory] = None,
+        scale_guide_loss: bool = False,
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec_with_llm/lightning.py:41]: Docstring for new arguments is missing.

File: replay/models/nn/sequential/sasrec_with_llm/model.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        schema: TensorSchema,
+        profile_emb_dim: int,
+        num_blocks: int = 2,
+        num_heads: int = 1,
+        hidden_size: int = 50,
+        max_len: int = 200,
+        dropout: float = 0.2,
+        ti_modification: bool = False,
+        time_span: int = 256,
+        reconstruction_layer: int = -1,
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec_with_llm/model.py:31]: Docstring for new arguments is missing

File: replay/models/nn/sequential/sasrec_with_llm/model.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            hidden_states = output_embeddings
+        else:
+            hidden_states = hidden_states__list[self.reconstruction_layer]
+        hidden_states_agg = self.weighting_fn(hidden_states, **self.weighting_kwargs)
+        return output_embeddings, hidden_states_agg
+
+    def aggregate_profile(self, user_profile_emb):
+        """
+        user_profile_emb: [batch_size, emb_dim]  или  [batch_size, K, emb_dim]
+        Возвращает: [batch_size, hidden_units] (если use_down_scale=True) либо [batch_size, emb_dim].
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec_with_llm/model.py]: Russian language

File: replay/models/nn/sequential/sasrec/model.py [hunk 8]
```diff
@@ -22,6 +22,7 @@ def __init__(
         dropout: float = 0.2,
         ti_modification: bool = False,
         time_span: int = 256,
+        **kwargs,
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec/model.py]: Do we need this kwargs here?

File: replay/models/nn/sequential/sasrec_with_llm/lightning.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+            hidden_state,
+            batch.null_profile_binary_mask_batch,
+            torch.nn.MSELoss(),
+        )
+        return self._merge_losses(loss_model, loss_guide)
+
+    def _merge_losses(self, loss_model: torch.Tensor, loss_guide: torch.Tensor) -> torch.Tensor:
+        epoch = self.trainer.current_epoch
+        if self.scale_guide_loss:
+            loss_model_val = loss_model.item()
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec_with_llm/lightning.py]: val means value, not validation? =)

File: replay/models/nn/sequential/sasrec/lightning.py [hunk 10]
```diff
@@ -69,7 +71,7 @@ def __init__(
         """
         super().__init__()
         self.save_hyperparameters()
-        self._model = SasRecModel(
+        self._model = sasrec_model_class(
```
Comment (antklen) [inline replay/models/nn/sequential/sasrec/lightning.py]: We suggest not to change SasRec class, as we want to make as few changes in the original code as possible.
Instead it is possible to delete self._model and re-initialize it with SasRecLLMModel inside SasRecLLM init. Of course it is not the best solution, but we prefer this way.

**Comment by shashist** [comment]
Thank you, my comments were resolved

File: replay/models/nn/sequential/sasrec/lightning.py (hunk 10 above)
Comment (shashist) [reply replay/models/nn/sequential/sasrec/lightning.py]: @Nikis14 Are deletion of self._model added before reinitialization?
